### PR TITLE
[SPARK-31348][SQL][DOCS] Document Join in SQL Reference

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Document join in SQL Reference.

### Why are the changes needed?
To make SQL Reference complete.

### Does this PR introduce any user-facing change?
Yes
Screen Shot 2020-04-05 at 8 46 47 PM

Screen Shot 2020-04-05 at 8 47 20 PM

Screen Shot 2020-04-05 at 8 47 41 PM

Screen Shot 2020-04-05 at 8 48 11 PM

Screen Shot 2020-04-05 at 8 48 33 PM

### How was this patch tested?
Manually build and check.

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @maropu Thanks!

File: docs/sql-ref-syntax-qry-select-join.md [hunk 1]
```diff
@@ -18,5 +18,52 @@ license: |
   See the License for the specific language governing permissions and
   limitations under the License.
 ---
+### Description
 
-**This page is under construction**
+A SQL join is used to combine rows from two relations based on join criteria. The following section describes the overall join syntax and the sub-sections cover different types of joins along with examples.
+
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-join.md:24]: We need a link to this section from the SELECT document, too?
https://github.com/apache/spark/blame/master/docs/sql-ref-syntax-qry-select.md#L78-L85

File: docs/sql-ref-syntax-qry-select-join.md [hunk 2]
```diff
@@ -18,5 +18,52 @@ license: |
...
+        | FULL [ OUTER ]
+        | [ LEFT ] ANTI
+      </code>
+  </dd>
+  <dt><code><em>join_criteria</em></code></dt>
+  <dd>
+    Specifies how the rows from one relation will be combined with the rows of another relation.<br><br>
+    <b>Syntax:</b>
+      <code>
+      ON booleanExpression | USING ( column_name [ , column_name ... ] )
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-join.md:57]: `booleanExpression` -> `boolean_expression`? Also, we need the corresponding explanation about `boolean_expression` here, too? https://github.com/apache/spark/blame/master/docs/sql-ref-syntax-qry-select.md#L42

File: docs/sql-ref-syntax-qry-select-join.md [hunk 3]
```diff
@@ -18,5 +18,52 @@ license: |
...
+</dl>
+
+### Join Types
+- [INNER JOIN](sql-ref-syntax-qry-select-join-inner.html)
+- [LEFT JOIN](sql-ref-syntax-qry-select-join-left.html)
+- [RIGHT JOIN](sql-ref-syntax-qry-select-join-right.html)
+- [FULL JOIN](sql-ref-syntax-qry-select-join-full.html)
+- [CROSS JOIN](sql-ref-syntax-qry-select-join-cross.html)
+- [SEMI JOIN](sql-ref-syntax-qry-select-join-semi.html)
+- [ANTI JOIN](sql-ref-syntax-qry-select-join-anti.html)
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-join.md:69]: We need these individual pages for the join types? I feel most of contents (e.g., `employee` and `department`) are overlapped there...

File: docs/sql-ref-syntax-qry-select-join.md (hunk 1 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-select-join.md:24]: It already has the ```Join relation``` in the ```from_item```. I will add a link to the ```Related Clauses``` section.

File: docs/sql-ref-syntax-qry-select-join.md (hunk 3 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-select-join.md:69]: I will change to one page and see how it looks.

Comment (maropu) [reply docs/sql-ref-syntax-qry-select-join.md:69]: Thanks!

File: docs/sql-ref-syntax-qry-select-join.md [hunk 4]
```diff
@@ -18,5 +18,222 @@ license: |
...
+  <dt><code><em>join_criteria</em></code></dt>
+  <dd>
+    Specifies how the rows from one relation will be combined with the rows of another relation.<br><br>
+    <b>Syntax:</b>
+      <code>
+      ON boolean_expression | USING ( column_name [ , column_name ... ] )
+      </code> <br><br>
+      <code>boolean_expression</code><br>
+      Specifies an expression with a return type of boolean.
+
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-join.md:61]: nit: remove the blank.

File: docs/sql-ref-syntax-qry-select-join.md [hunk 5]
```diff
@@ -18,5 +18,222 @@ license: |
...
+  </dd>
+</dl>
+
+### Join Types
+
+- <b>Inner Join </b><br>
+The inner join is the default join in Spark. It selects rows that have matching values in both relations.<br><br>
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-join.md:68]: nit: `Spark` -> `Spark  SQL`

Comment (maropu) [reply docs/sql-ref-syntax-qry-select-join.md:68]: nit: need intends

Comment (maropu) [reply docs/sql-ref-syntax-qry-select-join.md:68]: nit: need some indents

File: docs/sql-ref-syntax-qry-select-join.md [hunk 6]
```diff
@@ -18,5 +18,222 @@ license: |
...
+  |deptno |deptname   |
+  +-------+-----------+
+  |3      |Engineering|
+  |2      |Sales      |
+  |1      |Marketing  |
+  +-------+-----------+
+
+-- Use employee and department tables to demonstrate inner join.
+SELECT id, name, employee.deptno, deptname
+  FROM employee INNER JOIN department ON employee.deptno = department.deptno;
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-join.md:143]: nit: 2 space or 4 spaces in the head? I found 3 spaces... e.g., https://github.com/apache/spark/blame/master/docs/sql-ref-syntax-qry-select-groupby.md#L130

Comment (huaxingao) [reply docs/sql-ref-syntax-qry-select-join.md:143]: I prefer either 2 or 4 spaces. 
I checked a few SQL Reference:
DB2, Snowflake 2 spaces
SQL Server, PostgreSQL 4 spaces
Oracle 3 spaces

Comment (maropu) [reply docs/sql-ref-syntax-qry-select-join.md:156]: Ah, I see. Anything looks fine to me, but we need consistency in our SQL docs.

File: docs/sql-ref-syntax-qry-select-join.md [hunk 7]
```diff
@@ -18,5 +18,236 @@ license: |
...
+  </dd>
+  <dt><code><em>join_criteria</em></code></dt>
+  <dd>
+    Specifies how the rows from one relation will be combined with the rows of another relation.<br><br>
+    <b>Syntax:</b>
+      <code>
+      ON boolean_expression | USING ( column_name [ , column_name ... ] )
+      </code> <br><br>
+      <code>boolean_expression</code><br>
+      Specifies an expression with a return type of boolean.
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-join.md:60]: nit: the format?
```
  <dd>
    Specifies how the rows from one relation will be combined with the rows of another relation.<br><br>
    <b>Syntax:</b>
    <code>
      ON boolean_expression | USING ( column_name [ , column_name ... ] )
    </code> <br><br>
    <code>boolean_expression</code><br>
    Specifies an expression with a return type of boolean.
...
```

**Review by maropu — APPROVED** [review]
Looks fine except for the existing two comments.

**Comment by maropu** [comment]
cc: @srowen @viirya

File: docs/_data/menu-sql.yaml [hunk 8]
```diff
@@ -154,7 +154,9 @@
                   url: sql-ref-syntax-qry-select-distribute-by.html
                 - text: LIMIT Clause 
                   url: sql-ref-syntax-qry-select-limit.html
-                - text: Join Hints
+                - text: JOIN
+                  url: sql-ref-syntax-qry-select-join.html
+                - text: JOIN HINTS
```
Comment (viirya) [inline docs/_data/menu-sql.yaml:159]: Why we need to upper case of hints? We don't really have `HINTS` in query.

Comment (huaxingao) [reply docs/_data/menu-sql.yaml]: Fixed. Thanks!

Comment (huaxingao) [reply docs/_data/menu-sql.yaml]: I changed to ```JOIN hints``` last night, but actually ```JOIN``` is not a SQL keyword here and shouldn't be capitalized.  I will change it back to ```Join Hints```. Sorry for changing this back and forth.

File: docs/sql-ref-syntax-qry-select-join.md [hunk 9]
```diff
@@ -18,5 +18,236 @@ license: |
...
+### Description
 
-**This page is under construction**
+A SQL join is used to combine rows from two relations based on join criteria. The following section describes the overall join syntax and the sub-sections cover different types of joins along with examples.
+
+### Syntax
+{% highlight sql %}
+
+relation { [ join_type ] JOIN relation [ join_criteria ] | NATURAL join_type JOIN relation }
+
```
Comment (viirya) [inline docs/sql-ref-syntax-qry-select-join.md:29]: Not need to introduce NATURAL JOIN?

Comment (huaxingao) [reply docs/sql-ref-syntax-qry-select-join.md:29]: I am OK either way. 
@maropu WDYT?

Comment (maropu) [reply docs/sql-ref-syntax-qry-select-join.md:29]: Yea, better to document `NATURAL JOIN`, too, as the Pg doc does so: https://www.postgresql.org/docs/current/sql-select.html

**Comment by srowen** [comment]
Merged to master/3.0

**Comment by huaxingao** [comment]
Thanks, all!

**Pull request closed without merging**